File: py/mbug_dev/mbug_2820.py

from mbug_base import *

#==========================================================================
class mbug_2820(mbug):
    """ Represents an MBUG-2820 Environmeter device (Temperature & Humidity)"""

    _type = 2820

    def __init__(self, serial=None):
        mbug.__init__(self, self._type, serial)

    def read_raw(self):
        """ Get the current measurement raw data. """      
        self._write( (0xA0, 0x00, 0x00) )   # Read, raw format 
        d = self._read()
        raw = d[0]+(d[1]<<8)+(d[2]<<16)+(d[3]<<24)
        if (raw==0xffffffff): raise Exception('MBUG-2820 Sensor Error')
        return raw

    def from_raw(self, raw):
        if (raw==0xffffffff): raise Exception('MBUG-2820 Sensor Error')
        rH = ((raw>>16) & 0x3fff) * 100./2**14
        T = ((raw>>2) & 0x3fff) * 165./2**14 - 40
        return T, rH
        
    def read(self):
        """ Get the current measurement in physical units.
        Temperature in degrees Celsius, rel. humidity in percent. """
        self._write((0xA0, 0x00, 0x01))     # Read, ascii format, 1st part (temperature)
        s1 = self._read().strip('\0')
        self._write((0xA0, 0x01, 0x01))     # Read, ascii format, 2nd part (humidity)        
        s2 = self._read().strip('\0')
        try: T, rH = float(s1), float(s2)
        except:
            raise Exception('MBUG-2120 Device Error (Data: %s,%s)'%(s1,s2))
        if T<=-274 or rH<0: raise Exception('MBUG-2120 Sensor Error')
        return T, rH
    
#==========================================================================
def mbug_2820_test():
    # Print device list
    devs = list_devices(2820)
    print "Attached MBUG-2820 devices:"
    print devs if devs!=[] else 'None'

    # Open devices
    for ser in devs:
        try:
            print "Device", ser
            dev = mbug_2820(ser)
            ##data = dev.read_raw()
            ##T, rH = dev.from_raw(data)
            ##print "Data:", data, T, rH
            T,rH = dev.read()
            print "Data:", T, rH
        finally:
            try: dev.close()
            except: pass

#--------------------------------------------------------------------------    
if __name__ == "__main__":
   mbug_2820_test()
   pass

File: c/src/mbug_2820.c

#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include "mbug_2820.h"

//------------------------------------------------------------------------------
// Get a list of all available mbug-2820 devices.
const mbug_device_list  mbug_2820_list( void )
{
	return mbug_get_device_list(2820);
}

//------------------------------------------------------------------------------
// Init device to default parameters
static void mbug_2820_init( mbug_device dev )
{
	mbug_2820_set_acq_mode(dev, ACQ_MODE_INST);
}

//------------------------------------------------------------------------------
// Open a device specified by it's serial number (as int, 
// last digits of the serial number are matched only).
mbug_device mbug_2820_open( unsigned long serial_num )
{
	mbug_device dev = 
		mbug_open_int(2820, serial_num);
	if (!dev) return 0;

	mbug_2820_init( dev );

	return dev;
}

//------------------------------------------------------------------------------
// Open a device specified by it's id string, as returned
// by mbug_2820_list()
mbug_device mbug_2820_open_str( const char *id )
{
	mbug_device dev = 0;
	unsigned int type = mbug_type_from_id( id );
	if (type != 2820) return NULL;

	dev =  mbug_open_str( id );
	if (!dev) return 0;

	mbug_2820_init( dev );
	
	return dev;
}

//------------------------------------------------------------------------------
// Close a previously opened device.
void mbug_2820_close( mbug_device dev )
{
	mbug_close(dev);
}

//------------------------------------------------------------------------------
// Read raw data
long mbug_2820_read_raw( mbug_device dev )
{
	char data[8] = {0};
	int ret;

	ret = mbug_write( dev, "\xA0\x00\x00", 3);
	if (ret<0) return -1;

	ret = mbug_read( dev, data, 8 );
	if (ret<0) return -1;

	return *(long*)data;
}

//------------------------------------------------------------------------------
// Read temperatures and humidity. Pass a NULL pointer if you are not interested
// in the repective measurement value.
int mbug_2820_read( mbug_device dev, double *temperature, double *humidity )
{
	char data[16] = {0};
	int ret = 0;
	char* endp = 0;

	if (temperature != 0) *temperature = -300;
	if (humidity != 0) *humidity = -300;

	// Read 1st data packet
	ret = mbug_write( dev, "\xA0\x00\x01", 3);
	if (ret<0) return -1;
	ret = mbug_read( dev, data, 8 );
	if (ret<0) return -1;

	// Read 2nd data packet
	ret = mbug_write( dev, "\xA0\x01\x01", 3);
	if (ret<0) return -1;
	ret = mbug_read( dev, data+8, 8 );
	if (ret<0) return -1;

	// Convert temperature
	ret = 0;
	data[7] = '\0';
	if (temperature != 0) {
		errno = 0;
		*temperature = strtod(data, &endp);
		if (errno || *endp != '\0') {
			*temperature = -300;
			return -1;
		}
	}

	// Convert humidty
	data[15] = '\0';
	if (humidity != 0) {
		errno = 0;
		*humidity = strtod(data+8, &endp);
		if (errno || *endp != '\0') {
			*humidity = -300;
			return -1;
		}
	}

	// Test for device errors. Specific error code is located at data[5:6] as hex ascii.
	if (*temperature <= NOT_A_TEMPERATURE && data[4]=='\0' && data[5]!='\0') {
		*humidity = *temperature = -300 - strtol(data+5, 0, 16); // until an extended error handler is written: just add
		ret = -1;
	}

	return ret;
}


//------------------------------------------------------------------------------
// Read temperature only
double mbug_2820_read_temperature( mbug_device dev )
{
	double temperature = -300;
	mbug_2820_read(dev, &temperature, 0);
	return temperature;
}

//------------------------------------------------------------------------------
// Read humidity only
double mbug_2820_read_humidity( mbug_device dev )
{
	double humidity = -300;
	mbug_2820_read(dev, 0, &humidity);
	return humidity;
}

//------------------------------------------------------------------------------
// Set the acquisition mode (see mbug.h for documentation)
int mbug_2820_set_acq_mode( mbug_device dev, enum mbug_acquisition_mode mode )
{
	unsigned char cmd[2] = {0xF2, 0};
	switch (mode) {
		case ACQ_MODE_INST: cmd[1] = 0x00; break;
		case ACQ_MODE_LAST: cmd[1] = 0x01; break;
		case ACQ_MODE_NEXT: cmd[1] = 0x02; break;
		case ACQ_MODE_SYNC: cmd[1] = 0x03; break;
		default: // other modes not implemented
			return -1;
	}
	return mbug_write( dev, cmd, 2 );
}
//------------------------------------------------------------------------------
